src/dream.mli

```diff
@@ -696,7 +695,7 @@ val all_cookies : request -> (string * string) list
 
 (** {1 Bodies} *)
 
-val body : 'a message -> string promise
+val body : 'a message -> string Eio.Promise.or_exn
```

Comment:
I don't think we should expose a promise to the API user. We just need a promise at a high level internally inside this function so that concurrent calls to it resolve to the same body without causing concurrent reading of the same ephemeral stream.

Reply:
It's fine, I think, for concurrent calls to `body` to eventually return the same `string` in direct style, as long as internally it's implemented without racing on the internal streams.

Reply:
In other words, before this PR, the body promise was associated with the message, and exposed to the user. With Eio, we can still associate the promise with the message, but hide it from the user using an `await`. This PR originally disassociated the promise from the message, and created multiple promises, one for each concurrent call to `body`, on the call stack. This caused racing on the underlying body stream. I'm suggesting to again store a promise inside the message, as before this PR, and `await` on that promise at the top level inside `body`.

Reply:
s/on the call stack/on the call stack**S**